### PR TITLE
[MM-29157] Cloud invoices logic

### PR DESCRIPTION
**Pull request by ethervoid**
#### Summary

This PR includes all the need logic to get invoices or the invoice PDF for the installation 

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-29157

#### Related PR

https://github.com/mattermost/enterprise/pull/809

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: api4/cloud.go [hunk 1]
```diff
@@ -27,9 +31,12 @@ func (api *API) InitCloud() {
 	api.BaseRoutes.Cloud.Handle("/customer", api.ApiSessionRequired(getCloudCustomer)).Methods("GET")
 	api.BaseRoutes.Cloud.Handle("/customer", api.ApiSessionRequired(updateCloudCustomer)).Methods("PUT")
 	api.BaseRoutes.Cloud.Handle("/customer/address", api.ApiSessionRequired(updateCloudCustomerAddress)).Methods("PUT")
+	api.BaseRoutes.Cloud.Handle("/customer/address", api.ApiSessionRequired(updateCloudCustomerAddress)).Methods("PUT")
```
Comment (marianunez) [inline api4/cloud.go:34]: Is this change intended? Seems repeated from above

File: api4/cloud.go [hunk 2]
```diff
@@ -254,3 +261,68 @@ func confirmCustomerPayment(c *Context, w http.ResponseWriter, r *http.Request)
...
+		c.Err = model.NewAppError("Api4.getInvoicesForSubscription", "api.cloud.license_error", nil, "", http.StatusNotImplemented)
+		return
+	}
+
+	if !c.App.SessionHasPermissionTo(*c.App.Session(), model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	products, appErr := c.App.Cloud().GetInvoicesForSubscription()
```
Comment (marianunez) [inline api4/cloud.go:276]: Should this be called `invoices`?
```suggestion
	invoices, appErr := c.App.Cloud().GetInvoicesForSubscription()
```

File: web/context.go [hunk 3]
```diff
@@ -663,3 +663,15 @@ func (c *Context) RequireBotUserId() *Context {
 	}
 	return c
 }
+
+func (c *Context) RequireInvoiceId() *Context {
+	if c.Err != nil {
+		return c
+	}
+
+	if len(c.Params.InvoiceId) != 27 {
```
Comment (marianunez) [inline web/context.go:672]: Is this length guaranteed for all stripe invoice ids?

File: model/cloud.go [hunk 4]
```diff
@@ -84,3 +84,27 @@ type Subscription struct {
 	DNS        string   `json:"dns"`
 	IsPaidTier string   `json:"is_paid_tier"`
 }
+
+// Invoice model represents a cloud invoice
+type Invoice struct {
+	ID             string             `json:"id"`
+	Number         string             `json:"number"`
+	CreateAt       int64              `json:"create_at"`
+	Total          int64              `json:"total"`
```
Comment (marianunez) [inline model/cloud.go:93]: Are we missing the `Tax` field that we included in CWS?

**Review by marianunez — CHANGES_REQUESTED** [review]
Overall looks good. Some minor changes below:

File: web/context.go (hunk 3 above)
Comment (ethervoid) [reply web/context.go:672]: In the Stripe API is not set but by now is the length we have so we can verify its a valid id

File: model/cloud.go (hunk 4 above)
Comment (ethervoid) [reply model/cloud.go:93]: Added along with the Type for line item

File: api4/cloud.go (hunk 1 above)
Comment (ethervoid) [reply api4/cloud.go]: Nop, don't know how that was done but removing. Thanks!

**Review by marianunez — APPROVED** [review]

**Review by nickmisasi — APPROVED** [review]
